### PR TITLE
SRS Effect Models

### PR DESCRIPTION
**Pull request by agastya1208**
This PR improves and modifies the PR#132 by Emmanuel to incorporate the SRS models, and wavelength dependent fiber losses in Mininet Optical.
Currently, four SRS models have been included:
1. SRS Effect Model - _Computation taken from : M. Zirngibl Analytical model of Raman gain effects in massive wavelength division multiplexed transmission systems, 1998. - Equation 10. - Most accurate for uniform loading cases._
2. Zirngibl General Model - _The general form of the SRS model as proposed by M. Zirngibl in Analytical model of Raman gain effects in massive wavelength division multiplexed transmission systems, 1998." - Equation 7 - Most accurate for non-uniform loading cases._
3. Sylvestre SRS Model - _SRS model proposed in 'Raman-Induced Power Tilt in Arbitrarily Large Wavelength-Division-Multiplexed Systems' by T. Sylvestre et al., 2005 - Equation 6_
4. Bigo SRS Model - _SRS model proposed in the paper tilted: 'Experimental Investigation of Stimulated Raman Scattering Limitation on WDM Transmission Over Various Types of Fiber Infrastructures' by S.Bigo et al., 1999 - Equation 1 - Simplified form of the SRS_Effect_Model() function_

 - The SRS models scrips are in link.py file, and can be run by including in the Link() class, as part of keyword argument list. 
`from mnoptical.link import SRS_Effect_Model, Zirngibl_General_Model, Sylvestre_SRS_Model, Bigo_SRS_Model`
`...`
`...`
`srs_model = Zirngibl_General_Model`
`net = LinearTopology.build(power_dBm=power_dBm, ........., wdg_id='linear', srs_model=srs_model, wd_loss='SMF')`

- The 4 different tests earlier have been merged into 1 test - srs_tests.py, which checks all the models in one script. 
- The scripts _helper_funcs.py_ and _linear_params.py_ have also been modified to allow **params keyword arguments. 
- Wavelength dependent loss for fibers can also be specified with the keyword argument _wd_loss_ in Span() class. It takes two arguments for now - 'linear' for flat loss of 0.22dB/km; and 'SMF' for wavelength dependent loss spectra in _edfa_params.py_

### CODE REVIEW

**Review by lantz — CHANGES_REQUESTED** [review]
could you flatten this into a single commit and add Emmanuel as a co-author?

File: mnoptical/network.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def add_amplifier(self, name, amplifier_type='EDFA', **params):
         return amplifier
 
     def add_link(self, src_node, dst_node, src_out_port=-1,
-                 dst_in_port=-1, boost_amp=None, spans=None):
+                 dst_in_port=-1, boost_amp=None, spans=None, **params):
```
Comment (lantz) [inline mnoptical/network.py:73]: Probably should have been there anyway.

Comment (lantz) [reply mnoptical/network.py:73]: But, it should be documented in the docstring.

File: mnoptical/topo/helper_funcs.py [hunk 2]
```diff
@@ -4,11 +4,11 @@
 km = dB = dBm = 1.0
 m = .001
 
-def Span(km, amp=None):
+def Span(km, amp=None, **params):
```
Comment (lantz) [inline mnoptical/topo/helper_funcs.py:7]: Should be clear and documented that it's the  fiber span parameters.

File: mnoptical/topo/linear_params.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 class LinearTopology:
 
     @staticmethod
-    def build(power_dBm=0, span_length_km=3, span_no=1, hop_no=1, signal_no=10, debugger=False, wdg_id=None):
+    def build(power_dBm=0, span_length_km=3, span_no=1, hop_no=1, signal_no=10, debugger=False, wdg_id=None, **params):
```
Comment (lantz) [inline mnoptical/topo/linear_params.py:11]: It should be clear and documented that these are fiber/span parameters.

File: tests/srs_tests.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+  # Test for Zirngibl_General_Model
+  # M. Zirngibl Analytical model of Raman gain effects in massive wavelength division multiplexed transmission systems. - Equation 7
+  # Most accurate for non-uniform loading cases.
+
+  expected_result = 0.78
+  srs_model = Zirngibl_General_Model
+  preamp_tilt = srs_test(srs_model=srs_model)
+  assert preamp_tilt == expected_result
+
```
Comment (lantz) [inline tests/srs_tests.py:139]: Looks good.

**Review by lantz — CHANGES_REQUESTED** [review]
Looks good - the only missing bit is making sure you document what `**params` is. Also in the cases where it's not obvious what the params are for, perhaps it should be named appropriately? I suppose it's fairly obvious in `Span()` that it's the fiber span parameters. Another thing I'm not 100% sure is the parameter ordering - it used to be that  the positional parameter were (length, amp) which perhaps is nice. On the other hand, putting the span parameters together probably makes sense?

**Comment by lantz** [comment]
@agastya1208 Could you also add @manny118 as a coathor for this?

File: mnoptical/network.py (hunk 1 above)
Comment (agastya1208) [reply mnoptical/network.py:73]: Got it.. Do you mean, documented as in I should add a comment here clarifying that these are span parameters? I could change the argument name of **params in Span() to **span_params.. Does that sound good?

**Comment by agastya1208** [comment]
> @agastya1208 Could you also add @manny118 as a coathor for this?

I have already added `Co-authored` by: Emmanuel Akinrintoyo <emmanueljones2090@gmail.com>` in the commit message. I have not added a co-author this way before, Should I instead add the username @manny118 instead?

**Comment by lantz** [comment]
Double-checked coauthored-by and it looks good - not sure why it wasn't showing up before.

I think we're doing it wrong though. `srs_model` is called by `Span` so it really should be a feature and parameter of `Span` and not `Link`, don't you agree?

**Review by lantz — CHANGES_REQUESTED** [review]
This is close but I think the organization isn't quite right. If the SRS model is being called by `Span.propagate` then it is more logically a feature of `Span`. Then we don't need any extra parameters for `Link` etc..

**Comment by lantz** [comment]
No response from @agastya1208 but I guess we can merge this and fix it lateer.

**Pull request merged**